### PR TITLE
Rendering algorithm (vertex and primitive assembly)

### LINKED ISSUE #936: Re-organize GPURenderPipelineDescriptor according to the structure of the GPU pipeline
Talked about this on the editors meeting. We'd want to reorganize the pipeline descriptors according to the following guidelines:
  - a descriptor has a field per stage (be it programmable or fixed-point stage)
  - each field is a "stage descriptor" (e.g. output targets go into the "output merger", index related things go first, primitive topology is between the vertex shader and rasterization, etc)
  - fields are ordered according to the real order, in which they are specified on the GPUs (input assembly -> vertex -> primitive assembly -> rasterization -> fragment -> output merger)

The benefit is - consistency in what's inside a sub-dictionary versus just a field on the top-level descriptor, and clear understanding on *when* things happen on the GPU.

### PR DESCRIPTION
**Pull request by kvark**
Closes #936 
Based on #1288

First stab at describing how render pipeline works.

***
<a href="https://pr-preview.s3.amazonaws.com/kvark/gpuweb/pull/1293.html" title="Last updated on Jan 28, 2021, 4:36 PM UTC (093e7df)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/1293/0e7b94e...kvark:093e7df.html" title="Last updated on Jan 28, 2021, 4:36 PM UTC (093e7df)">Diff</a>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -3693,21 +3693,111 @@ GPURenderPipeline includes GPUPipelineBase;
...
+            - |indexBuffer|: Currently bound index buffer via {{GPURenderEncoderBase/setIndexBuffer()}}.
+            - |vertexBuffers|: [=list=]&lt;vertex buffer&gt; bound by {{GPURenderEncoderBase/setVertexBuffer()}}.
+            - |bindGroups|: [=list=]&lt;{{GPUBindGroup}}&gt; bound by {{GPUProgrammablePassEncoder/setBindGroup(index, bindGroup, dynamicOffsets)}}.
+
+        Issue: define somewhere what a render pass state is?
+        Let |vertexState| be |descriptor|.{{GPURenderPipelineDescriptor/vertexState}}.
+
+        1. Resolve indices.
+            The pipeline builds a list of vertices to process for each instance.
+            If |vertexState|.{{GPUVertexStateDescriptor/indexFormat}} is not `null`:
```
Comment (Kangz) [inline spec/index.bs:3723]: I think we want to check if the drawcall is indexed or not here instead of the indexFormat

File: spec/index.bs [hunk 2]
```diff
@@ -3693,21 +3693,111 @@ GPURenderPipeline includes GPUPipelineBase;
...
+                            2 bytes of fixed-point unsigned 8-bit integers into 2 floating-point values
+                            as `vec2<f32>` in WGSL.
+                        </div>
+                    1. Bind the |data| to vertex shader input
+                        location |attributeDesc|.{{GPUVertexAttributeDescriptor/shaderLocation}}.
+            1. For each {{GPUBindGroup}} group at |index| in |bindGroups|:
+                1. Bind the group of resources visible to vertex shader group |index|.
+            1. Invoke vertex shader in |descriptor|.{{GPURenderPipelineDescriptor/vertexStage}}.
+
+                Note: The target platform caches the results of vertex shader invocations.
```
Comment (Kangz) [inline spec/index.bs:3762]: And similarly a vertexIndex may be processed more times than it is listed, for example when breaking up long strips.

File: spec/index.bs [hunk 3]
```diff
@@ -3693,21 +3693,111 @@ GPURenderPipeline includes GPUPipelineBase;
...
+                        <div class="example">
+                        An attribute of type {{GPUVertexFormat/"uchar2norm"}} will be converted from
+                            2 bytes of fixed-point unsigned 8-bit integers into 2 floating-point values
+                            as `vec2<f32>` in WGSL.
+                        </div>
+                    1. Bind the |data| to vertex shader input
+                        location |attributeDesc|.{{GPUVertexAttributeDescriptor/shaderLocation}}.
+            1. For each {{GPUBindGroup}} group at |index| in |bindGroups|:
+                1. Bind the group of resources visible to vertex shader group |index|.
+            1. Invoke vertex shader in |descriptor|.{{GPURenderPipelineDescriptor/vertexStage}}.
```
Comment (Kangz) [inline spec/index.bs:3760]: We might even model vertex shaders as functions of bindings, inputs(builtins and attributes) that returns outputs (varyings and builtins) while potentially doing side effects. Then call that explicitly.

The other thing is that we should be explicit as to which steps happen in parallel. Maybe we could say that the function executes asynchonously. And in general we should explain what parts of rendering happen on one another, and which parts are done in parallel. (simlarly we could describe that with the concept of promises maybe?)

File: spec/index.bs (hunk 1 above)
Comment (kvark) [reply spec/index.bs:3723]: good point!

File: spec/index.bs (hunk 3 above)
Comment (kvark) [reply spec/index.bs:3760]: At some point, we need to say that we give the driver control to invoke our shader. Whether here, or inside another algorithm. Is there a benefit for wrapping this "invoke xxx" inside a separate algorithm?

Comment (Kangz) [reply spec/index.bs:3767]: I'm not sure what's best. What I feel stronger about is that the synchronization guarantees and the parallel nature of GPUs should be visible in the spec. (like VS and FS for a render pass running in parallel or out of order for different triangles)

Comment (kvark) [reply spec/index.bs:3767]: Yes, I totally agree. The PR now has a bit of wording to be clear about that. We'll add more.

File: spec/index.bs [hunk 4]
```diff
@@ -3686,21 +3693,162 @@ GPURenderPipeline includes GPUPipelineBase;
 
 <script type=idl>
 dictionary GPURenderPipelineDescriptor : GPUPipelineDescriptorBase {
+    GPUVertexStateDescriptor vertexState = {};
     required GPUProgrammableStageDescriptor vertexStage;
-    GPUProgrammableStageDescriptor fragmentStage;
-
     required GPUPrimitiveTopology primitiveTopology;
```
Comment (kainino0x) [inline spec/index.bs:3698]: probably index format should be in the primitive assembly stage. see #767

Comment (kvark) [reply spec/index.bs:3698]: that's what #1307 does, which we should focus on before continuing on this one

Comment (kainino0x) [reply spec/index.bs:3698]: I just recorded this as a quick note during the editing meeting just to make sure we didn't forget about it when you created #1307. So marking as resolved now.

**Comment by Kangz** [comment]
It might be easier for iterations to make an issue / PR with just the IDL for things so that you don't need to update all the Bikeshed references.

**Comment by kvark** [comment]
This is now rebased, doesn't contain any API changes, and can be merged. More stuff to follow, of course, but we don't have to wait for it.

File: spec/index.bs [hunk 5]
```diff
@@ -3835,6 +3835,153 @@ dictionary GPURenderPipelineDescriptor : GPUPipelineDescriptorBase {
...
+                - assign the |vertexList| item |i| to be |drawCall|.firstVertex + |i|
+
+            Note: in case of indirect draw calls, the `indexCount`, `vertexCount`,
+            and other properties of |drawCall| are read from the indirect buffer
+            instead of the draw command itself.
+
+            Issue: specify indirect commands better.
+
+        1. Process vertices. Each vertex |vertexIndex| in the |vertexList|,
+            in each instance of index |rawInstanceIndex|, is processed independenty.
```
Comment (kainino0x) [inline spec/index.bs:3868]:
```suggestion
            in each instance of index |rawInstanceIndex|, is processed independently.
```

File: spec/index.bs [hunk 6]
```diff
@@ -3835,6 +3835,153 @@ dictionary GPURenderPipelineDescriptor : GPUPipelineDescriptorBase {
...
+            The |rawInstanceIndex| is in range from 0 to |drawCall|.instanceCount - 1, inclusive.
+            This |processing happens in parallel, and any side effects, such as
```
Comment (kainino0x) [inline spec/index.bs:3870]:
```suggestion
            This processing happens in parallel, and any side effects, such as
```

File: spec/index.bs [hunk 7]
```diff
@@ -3835,6 +3835,153 @@ dictionary GPURenderPipelineDescriptor : GPUPipelineDescriptorBase {
...
+            - |vertexBuffers|: [=list=]&lt;vertex buffer&gt; bound by {{GPURenderEncoderBase/setVertexBuffer()}}.
+            - |bindGroups|: [=list=]&lt;{{GPUBindGroup}}&gt; bound by {{GPUProgrammablePassEncoder/setBindGroup(index, bindGroup, dynamicOffsets)}}.
+
+        Issue: define somewhere what a render pass state is?
+
+        1. Resolve indices.
+            The pipeline builds a list of vertices to process for each instance.
+            If |drawCall| is an indexed draw call:
+                - initialize the |vertexList| with |drawCall|.indexCount integers
+                - assign the |vertexList| item |i| to be [$fetch index$](|i| + |drawCall|.firstIndex, |indexBuffer|.buffer, |indexBuffer|.offset, |indexBuffer|.format) + |drawCall|.baseVertex
```
Comment (kainino0x) [inline spec/index.bs:3856]: slightly unclear that `|i|` is being defined here.

File: spec/index.bs [hunk 8]
```diff
@@ -3835,6 +3835,153 @@ dictionary GPURenderPipelineDescriptor : GPUPipelineDescriptorBase {
...
+            Issue: specify indirect commands better.
+
+        1. Process vertices. Each vertex |vertexIndex| in the |vertexList|,
+            in each instance of index |rawInstanceIndex|, is processed independenty.
+            The |rawInstanceIndex| is in range from 0 to |drawCall|.instanceCount - 1, inclusive.
+            This |processing happens in parallel, and any side effects, such as
+            writes into {{GPUBufferBindingType/"storage"|GPUBufferBindingType."storage"}} bindings,
+            may happen in any order.
+            1. Let |instanceIndex| be |rawInstanceIndex| + |drawCall|.baseInstance.
+            1. For each non-`null` |vertexBufferLayout| under index |i|
```
Comment (kainino0x) [inline spec/index.bs:3874]: unclear that `|i|` is being defined here (I think it is?)

File: spec/index.bs [hunk 9]
```diff
@@ -3835,6 +3835,153 @@ dictionary GPURenderPipelineDescriptor : GPUPipelineDescriptorBase {
...
+                    </dl>
+                1. For each |attributeDesc| in |vertexBufferLayout|.{{GPUVertexBufferLayout/attributes}}:
+                    1. Let |attributeOffset| be |vertexBufferOffset| +
+                        |vertexElementIndex| * |vertexBufferLayout|.{{GPUVertexBufferLayout/arrayStride}} +
+                        |attributeDesc|.{{GPUVertexAttribute/offset}}.
+                    1. Load the attribute |data| of format |attributeDesc|.{{GPUVertexAttribute/format}}
+                        from |vertexBuffer| starting at offset |attributeOffset|.
+                    1. Convert the |data| into a shader-visible format.
+                        <div class="example">
+                        An attribute of type {{GPUVertexFormat/"uchar2norm"}} will be converted from
```
Comment (kainino0x) [inline spec/index.bs:3892]:
```suggestion
                            An attribute of type {{GPUVertexFormat/"uchar2norm"}} will be converted from
```

File: spec/index.bs [hunk 10]
```diff
@@ -3835,6 +3835,153 @@ dictionary GPURenderPipelineDescriptor : GPUPipelineDescriptorBase {
...
+                            2 bytes of fixed-point unsigned 8-bit integers into 2 floating-point values
+                            as `vec2<f32>` in WGSL.
+                        </div>
+                    1. Bind the |data| to vertex shader input
+                        location |attributeDesc|.{{GPUVertexAttribute/shaderLocation}}.
+            1. For each {{GPUBindGroup}} group at |index| in |bindGroups|:
```
Comment (kainino0x) [inline spec/index.bs:3898]: Perhaps an inner loop over the group members instead of "binding" the whole group?

File: spec/index.bs [hunk 11]
```diff
@@ -3835,6 +3835,153 @@ dictionary GPURenderPipelineDescriptor : GPUPipelineDescriptorBase {
...
+                1. Bind the group of resources visible to vertex shader group |index|.
+            1. Invoke vertex shader in |descriptor|.{{GPURenderPipelineDescriptor/vertex}}.{{GPUProgrammableStage}}.
```
Comment (kainino0x) [inline spec/index.bs:3900]:
```suggestion
            1. Invoke vertex entry point with name
                |descriptor|.{{GPURenderPipelineDescriptor/vertex}}.{{GPUProgrammableStage/entryPoint}} in
                |descriptor|.{{GPURenderPipelineDescriptor/vertex}}.{{GPUProgrammableStage/module}}.
```

File: spec/index.bs [hunk 12]
```diff
@@ -3835,6 +3835,153 @@ dictionary GPURenderPipelineDescriptor : GPUPipelineDescriptorBase {
...
+
+                Note: The target platform caches the results of vertex shader invocations.
+                There is no guarantee that any |vertexIndex| that repeats more than once will
+                result in multiple invocations. Similarly, no guarantee that a single |vertexIndex|
```
Comment (kainino0x) [inline spec/index.bs:3904]:
```suggestion
                result in multiple invocations. Similarly, there is no guarantee that a single |vertexIndex|
```

File: spec/index.bs [hunk 13]
```diff
@@ -3835,6 +3835,153 @@ dictionary GPURenderPipelineDescriptor : GPUPipelineDescriptorBase {
...
+                will only be processed once.
+
+                1. Set the `VertexIndex` builtin to |vertexIndex|.
+                1. Set the `InstanceIndex` builtin to |instanceIndex|.
```
Comment (kainino0x) [inline spec/index.bs:3908]: "if any"?
```suggestion
                1. Set the `VertexIndex` builtin, if any, to |vertexIndex|.
                1. Set the `InstanceIndex` builtin, if any, to |instanceIndex|.
```

File: spec/index.bs [hunk 14]
```diff
@@ -3835,6 +3835,153 @@ dictionary GPURenderPipelineDescriptor : GPUPipelineDescriptorBase {
...
+        1. Assemble primitives.
+            For each instance, the primitives get assembled from the vertices that have been
+            processed by the shaders, based on the |vertexList|.
+          
+            1. First, if |descriptor|.{{GPURenderPipelineDescriptor/primitive}}.{{GPUPrimitiveState/stripIndexFormat}} is not `null`
+                (which means the primitive topology is a strip), and the |drawCall| is indexed,
+                the |vertexList| is split into sub-lists
+                using the maximum value of this index format as a separator.
+
+                Example: a |vertexList| with values `[1,2,65535,4,5,6]` of type {{GPUIndexFormat/"uint16"}}
```
Comment (kainino0x) [inline spec/index.bs:3920]:
```suggestion
                Example: a |vertexList| with values `[1, 2, 65535, 4, 5, 6]` of type {{GPUIndexFormat/"uint16"}}
```

**Review by kainino0x — APPROVED** [review]

File: spec/index.bs [hunk 15]
```diff
@@ -3788,7 +3788,7 @@ Render [=pipeline=] outputs are:
   - the color attachments, described by {{GPUColorTargetState}}
   - optionally, depth-stencil attachment, described by {{GPUDepthStencilState}}
 
-Stages of a render [=pipeline=]:
+<dfn dfn>Render stages</dfn> of a [=pipeline=]:
```
[the remark this answers is not in the document]
Comment (kvark) [reply spec/index.bs:3791]: This was actually intentional. If you look at Vulkan pipeline description, it's one big pipeline, and it has both compute and render stages in it.

File: spec/index.bs (hunk 11 above)
Comment (kvark) [reply spec/index.bs:3905]: That's too verbose, I think. Any reason we can't refer to the whole `GPUProgrammableStage` here?

**Review by kvark — COMMENTED** [review]
Nicely detailed review, thanks @kainino0x !
All except one points are addressed.

File: spec/index.bs (hunk 11 above)
Comment (kainino0x) [reply spec/index.bs:3905]: That would be fine, in that case
```suggestion
            1. Invoke vertex shader entry point described by |descriptor|.{{GPURenderPipelineDescriptor/vertex}}.
```

File: spec/index.bs [hunk 16]
```diff
@@ -3889,23 +3891,26 @@ configuring each of the [=render stages=].
...
                             2 bytes of fixed-point unsigned 8-bit integers into 2 floating-point values
                             as `vec2<f32>` in WGSL.
                         </div>
                     1. Bind the |data| to vertex shader input
                         location |attributeDesc|.{{GPUVertexAttribute/shaderLocation}}.
             1. For each {{GPUBindGroup}} group at |index| in |bindGroups|:
-                1. Bind the group of resources visible to vertex shader group |index|.
+                1. For each resource {{GPUBindingResource}} in the bind group:
+                    1. Let |layout| be the corresponding {{GPUBindGroupLayoutEntry}} for this resource.
+                    1. If |layout|.{{GPUBindGroupLayoutEntry}}.visibility includes {{GPUShaderStage/VERTEX}}:
```
Comment (kainino0x) [inline spec/index.bs:3903]:
```suggestion
                    1. If |layout|.{{GPUBindGroupLayoutEntry/visibility}} includes {{GPUShaderStage/VERTEX}}:
```

also maybe `|layout|` should be `|entry|`?

File: spec/index.bs [hunk 17]
```diff
@@ -3954,7 +3959,9 @@ configuring each of the [=render stages=].
                 Any incomplete primitives are dropped.
 
         1. Rasterize.
-            Issue: fill out the section
+            Rasterization is a hardware processing stage that maps the
+            generated primitives to the actual screen.
+            1. Apply {{GPUPrimitiveState/cullMode}}
```
Comment (kainino0x) [inline spec/index.bs:3964]: ?

**Review by kainino0x — APPROVED** [review]

File: spec/index.bs (hunk 11 above)
Comment (toji) [reply spec/index.bs:3905]: I like Kai's edit here.

File: spec/index.bs [hunk 18]
```diff
@@ -3835,6 +3835,160 @@ dictionary GPURenderPipelineDescriptor : GPUPipelineDescriptorBase {
...
+                    1. If |layout|.{{GPUBindGroupLayoutEntry}}.visibility includes {{GPUShaderStage/VERTEX}}:
+                        - Bind the resource to the shader under group |index| and binding {{GPUBindGroupLayoutEntry/binding|GPUBindGroupLayoutEntry.binding}}.
+            1. Invoke vertex shader in |descriptor|.{{GPURenderPipelineDescriptor/vertex}}.{{GPUProgrammableStage}}.
+
+                Note: The target platform caches the results of vertex shader invocations.
+                There is no guarantee that any |vertexIndex| that repeats more than once will
+                result in multiple invocations. Similarly, there is no guarantee that a single |vertexIndex|
+                will only be processed once.
+
+                1. Set the `VertexIndex` builtin, if any, to |vertexIndex|.
```
Comment (toji) [inline spec/index.bs:3912]: Shouldn't setting the builtins come before invoking the shader?

File: spec/index.bs (hunk 17 above)
Comment (toji) [reply spec/index.bs:3964]: Yeah, this feels like it got pasted into the wrong spot? Culling happens during primitive assembly. Relatedly, there should probably be a brief mention of primitives being culled due to sitting fully outside the normalized device coordinates.

**Review by toji — APPROVED** [review]

**Pull request merged**